Add documentation on broadcasting (#3356)
PennyLaneAI/pennylane: doc/introduction/circuits.rst

constraints:

* The quantum function accepts classical inputs, and consists of

>   :doc:`quantum operators <operations>` or sequences of operators called :doc:`templates`.

Author: Instructions like
```python
[qml.RX(x, wires=w) for w in dev.wires]
```
are allowed.

* The function can contain classical flow control structures such as ``for`` loops or ``if`` statements.

* The quantum function must always return either a single or a tuple of
  *measurement values*, by applying a :doc:`measurement function <measurements>`
  to the qubit register. The most common example is to measure the expectation value of
  a :ref:`qubit observable <intro_ref_ops_qobs>` or

>   :ref:`continuous-value observable <intro_ref_ops_cvobs>`.

Author: Not only observable measurements are supported by PennyLane. In particular, `qml.var` includes post-processing of observable measurements.

.. note::

>     Quantum functions are evaluated on a device from within a QNode.

Author: They typically are executed via QNodes, and it is the user-facing way to do it. But it is not the only one.

Wires
*****

> The wires argument can be an integer that defines the *number of wires*

Author: This "either...or" construction was not closed.

measurement statistics *exactly*.

Note that this argument can be temporarily overwritten when a QNode is called. For example, ``my_qnode(shots=3)``
will temporarily evaluate ``my_qnode`` using three shots. This is a feature of each QNode and it is not 

> necessary to manually implement the ``shots`` keyword argument of the quantum function.

Author: For clarification.

    result = circuit(0.543)

Parameter Broadcasting in QNodes
--------------------------------

> Some :class:`~.pennylane.QNode` circuits support the execution at multiple parameters
> simultaneously:

Reviewer: Maybe?
```suggestion
Depending on the quantum operations used, a :class:`~.pennylane.QNode` may support execution at multiple parameters simultaneously:
```

makes it less ambiguous wondering what causes the lack of support

>>> x = np.array([0.543, 1.234])
>>> result = circuit(x)
>>> result
tensor([0.85616242, 0.33046511], requires_grad=True)

> This is called *parameter broadcasting* (as for, say, NumPy functions executed along an axis)
> or *parameter batching* (as in the application of a function to a *batch* of parameters in
> machine learning).

Reviewer: ```suggestion
Note that we are passing in a 1-dimensional array of parameters to the `circuit()`
QNode defined above, which takes a single parameter and returns a single expectation
value. As the input is now an array, the output is also an array, with each
element the expectation value of the corresponding input element.

This is called *parameter broadcasting* (as for, say, NumPy functions executed along an axis)
or *parameter batching* (as in the application of a function to a *batch* of parameters in
machine learning).
```

Author: Nice! :)

In addition to a more flexible execution syntax, broadcasting can yield performance boosts
compared to the separate execution of the QNode for each parameter setting. Whether or not this is
the case depends on quite a few details, but in particular for (at most) moderately sized circuits
(:math:`\lesssim 20` wires) with a moderate number of parameters (:math:`\lesssim 200`) executed
on a classical simulator, one can expect to benefit from broadcasting.
See the :class:`~.pennylane.QNode` documentation for usage details.

> Many standard quantum operators support broadcasting, see the corresponding attribute

Reviewer: ```suggestion
Many standard quantum operators support broadcasting; see the corresponding attribute
```

:obj:`~.pennylane.ops.qubit.attributes.supports_broadcasting` for a list. The
:class:`~.pennylane.operation.Operator` documentation contains implementation details
and a guide to make custom operators compatible with broadcasting.

> Broadcasting can be used with any device, but will usually only yield performance upgrades for
> classical simulator devices like ``"default.qubit"`` that indicate that they support it:

Reviewer: ```suggestion
Broadcasting can be used with any device, but will usually only yield performance upgrades for
devices like ``"default.qubit"`` that indicate that they support it:
```

I have a feeling that even non-simulator devices will see speedups, partly due to avoiding slow python for loops.

Author: Ah, yes, in the QNode documentation I mentioned this but here I do not explicitly mention it, so indeed it's good to not restrict the statement, I think :)
